pkg/plugin/internal/dashboard/errors.go

```diff
@@ -0,0 +1,8 @@
+package dashboard
```

Comment:
I dont really see the value of this file. Either we define all the custom errors used in the package here or remove this file and use `fmt.Errorf` as we are doing everywhere.

Reply:
Errors, produced inside a package, should be defined as package level variable inside the package. That part of a well designed error handing which exists since Go 1.13

https://go.dev/blog/go1.13-errors#errors-and-package-apis

There is also a linter for that, included in golangci-lint: https://github.com/Djarvur/go-err113

Reply:
I understand that design pattern for errors that are reused in the package. But the errors you defined is only used once. Then the question is why defining just these errors? There are plenty of custom `fmt.Errorf()` in package and we should move all of them into `errors.go`. I guess the linter suggests to replace `fmt.Errorf()` with `errors.New()`. But I prefer to keep using `fmt.Errorf()` formating the error is straight forward where as we need to use `fmt.Sprintf()` inside `errors.New()`.

As each function returns its own custom error, I think its better to keep it close to the function. If we are reusing the same error atleast twice then it justifies definining it in new `errors.go`. Does it make sense?

Reply:
> Then the question is why defining just these errors

That current error handling design in the go language. Error, that coming from that packages, should be created as package level initiated error. Error, coming from external packages, should be wrapped with `fmt.Errorf` together with the `%w` placeholder.

The goal is that caller can be used `errors.Is` to check for an specific error. Before go 1.13, string compare where is to indicate specific errors. In tests, I could also check for specific errors.

https://github.com/jkroepke/grafana-dashboard-reporter-app/blob/a7a685612010902d854c6da7157caa06d891143b/pkg/plugin/client/client_test.go#L89

Basic each guide to error handling leads to that current implementation.

> There are plenty of custom fmt.Errorf()

This is not technically correct. I go through most of the error and we are not creating any new custom errors. Instead, with `fmt.Errorf`, we wrap returned errors from external packages. The different is `%w` on the errors, we can't be inside an errors.New declaration. 

Please do not mix creating/define new errors and wrap errors.

> I think its better to keep it close to the function.

That is still the case:

for example:

```go
	if len(dashboardData) == 0 {
		return nil, ErrJavaScriptReturnedNoData
	}
```

It's visible, which error is returned. It required to give package level errors verbose names.

---

What may could help to understand that concept:

with `var ErrJavaScriptReturnedNoData = errors.New("javascript did not return any dashboard data")`, it's the go-away to defined new Exception Types and in Java or JavaScript, you may also define multiple custom exceptions for different scenarios. While in java `try {} catch {ErrJavaScriptReturnedNoData e}` exists, in go you just using `if errors.Is(err, ErrJavaScriptReturnedNoData)`

---

Additional resource:

- https://tech.new-work.se/designing-errors-with-go-1-13-81692537457
- https://go.dev/blog/go1.13-errors

Example for go stdlib:

- https://github.com/golang/go/blob/1f0c044d60211e435dc58844127544dd3ecb6a41/src/flag/flag.go#L99-L109
- https://github.com/golang/go/blob/1f0c044d60211e435dc58844127544dd3ecb6a41/src/net/interface.go#L31-L37
- https://pkg.go.dev/io#ErrClosedPipe

Reply:
Ahhh, I was not thinking about `errors.Is()`. Yes, I agree with you in that case, they deserve to be declared. Thanks for the detailed explanation. Appreciate it!